metrics-core/src/main/java/com/yammer/metrics/stats/ExponentiallyDecayingSample.java

```diff
@@ -164,12 +174,15 @@ private void rescale(long now, long next) {
         if (nextScaleTime.compareAndSet(next, now + RESCALE_THRESHOLD)) {
             lockForRescale();
             try {
-                final long oldStartTime = startTime;
-                this.startTime = currentTimeInSeconds();
+                final int oldOffset = startOffset;
+                this.startOffset = currentOffset();
                 final ArrayList<Double> keys = new ArrayList<Double>(values.keySet());
                 for (Double key : keys) {
-                    final Long value = values.remove(key);
-                    values.put(key * exp(-alpha * (startTime - oldStartTime)), value);
+                    final ValueWithOffset value = values.remove(key);
+                    int age = startOffset - value.getOffset();
+                    if (age < maxAge) {
```

Comment:
What we seem to be trying to do with this change is to create a forward decaying [1] _time windowed_ priority reservoir [2]. In the presence of few samples (say, less than 1028 per `max age` window), the reported values will be perfectly accurate (since we will include all of them).  In the presence of many samples, the reported values will represent a sampling over the last `max age` period. In this case, it will be as if we had started the histogram `max age` ago.

Here are the invariants I'd like us to be able to offer:

1. Any data point is represented in a bounded number of reporting windows (preferably 1, i.e. any bad event should be reported exactly once)
2. Any "bad" event is represented in at least one reporting window
3. With low throughput comes increased accuracy
4. A bad 99th from a histogram indicates an active problem

What I really like about this change is that we get the following:

1. Any data point _is_ represented in a bounded number of reporting windows (ratio of reporting period to max age)
2. Any bad event is _more likely_ to be represented in at least one window, since we're evicting old entries, which cause more current entries to make it to the sample
3. Decreasing throughput _will_ increase accuracy, up to perfect accuracy at fewer than 1028 samples per `max age` period
4. A bad 99th _will_ indicate an active problem (or at least, one that is no older than `max age`)

I think my initial suggestion for `max age` would be twice the reporting interval?

[1]: [Forward Decay: A Practical Time Decay Model for Streaming Systems](https://pdfs.semanticscholar.org/1bbf/70755a1716b283fec3a747d1d978faa64bb8.pdf)
[2]: [Estimating arbitrary subset sums with few probes](http://www.tau.ac.il/~nogaa/PDFS/adltproc1.pdf)